File: project/lakehouse/silver/src/main/java/csx55/sta/silver/jobs/EventsCleanedJob.java
package csx55.sta.silver.jobs;

import csx55.sta.streaming.config.StreamConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Silver Layer Job: Events Cleaning
 *
 * <p><b>Purpose:</b> Data quality and validation for special events data (contextual data).</p>
 *
 * <p><b>SQL-FIRST PATTERN DEMONSTRATION:</b></p>
 * <p>This class is intentionally minimal (only 20 lines of actual code).
 * <b>All business logic lives in SQL</b>: {@code sql/events_cleaned.sql}</p>
 *
 * <p><b>LOAN ORIGINATION PARALLEL:</b></p>
 * <pre>
 * Special Events Data            → Loan Special Circumstances
 * ─────────────────────────────────────────────────────────────────
 * Attendance validation          → Number of co-borrowers validation (1-4)
 * Start/end time validation      → Loan term date range validation
 * Event type enum                → Loan program type (FHA, VA, Conventional)
 * Event duration check           → Loan term validation (6mo - 30yr)
 * Major event flag               → Has guarantor flag (risk reduction)
 * Venue name (optional)          → Guarantor name (optional)
 *
 * Teaching Point: Special circumstances need their own validation rules.
 * Not all data follows the same pattern - some is optional but still needs validation.
 * </pre>
 *
 * <p><b>What This Job Does:</b></p>
 * <ul>
 *   <li>Reads from {@code lakehouse.bronze.events} (raw special events data)</li>
 *   <li>Applies SQL validation rules (attendance, time ranges, event types)</li>
 *   <li>Validates logical consistency (start before end, reasonable duration)</li>
 *   <li>Adds data quality flags and derived event impact metrics</li>
 *   <li>Writes to {@code lakehouse.silver.events_cleaned} (validated events)</li>
 * </ul>
 *
 * @see AbstractSilverJob
 */
public class EventsCleanedJob extends AbstractSilverJob {

    private static final Logger logger = LoggerFactory.getLogger(EventsCleanedJob.class);

    /**
     * Constructor.
     * Initializes with configuration for events cleaning job.
     *
     * @param config Global streaming configuration
     */
    public EventsCleanedJob(StreamConfig config) {
        super(config, config.getSilverEventsCleanedConfig());
    }

    /**
     * Specifies which SQL file contains the business logic.
     *
     * <p><b>Teaching Note:</b> This is the ONLY line that differs between
     * WeatherCleanedJob, EventsCleanedJob, and TripsCleanedJob.
     * Everything else is identical - that's template reusability!</p>
     *
     * @return Path to SQL file (relative to resources directory)
     */
    @Override
    protected String getSqlFilePath() {
        return "sql/events_cleaned.sql";
    }

    /**
     * Main entry point for the job.
     *
     * @param args Command-line arguments (not used)
     */
    public static void main(String[] args) {
        logger.info("========================================");
        logger.info("Silver Layer: Events Cleaned Job");
        logger.info("========================================");

        try {
            // Load configuration
            StreamConfig config = new StreamConfig();

            // Create and run job
            EventsCleanedJob job = new EventsCleanedJob(config);
            job.run();

        } catch (Exception e) {
            logger.error("Failed to run EventsCleanedJob", e);
            System.exit(1);
        }
    }
}

File: project/lakehouse/silver/src/main/scala/csx55/sta/silver/jobs/EventsCleanedJob.scala
package csx55.sta.silver.jobs

import com.amazon.deequ.checks.{Check, CheckLevel}
import csx55.sta.silver.config.BusinessRules
import csx55.sta.streaming.config.StreamConfig
import org.slf4j.LoggerFactory

class EventsCleanedJob(config: StreamConfig)
    extends AbstractSilverJob(config, config.getSilverEventsCleanedConfig()) {

  override protected def getSqlFilePath(): String = {
    "sql/events_cleaned.sql"
  }

  override protected def getDeequChecks(): Check = {
    Check(CheckLevel.Error, "Events Data Quality Checks")
      .isComplete("timestamp")
      .isComplete("location_id")
      .isComplete("event_type")    // SPORTS, CONCERT, CONFERENCE, FESTIVAL
      .isComplete("start_time")    // Event start timestamp
      .isComplete("end_time")      // Event end timestamp
      .hasMin("attendance_estimate", _ >= 1.0)
      .hasMax("attendance_estimate", _ <= 100000.0)
  }

  override protected def getBusinessRuleReplacements(): Map[String, Any] = {
    Map(
      // Major event identification thresholds
      "MAJOR_EVENT_ATTENDANCE_THRESHOLD" -> BusinessRules.MAJOR_EVENT_ATTENDANCE_THRESHOLD,
      "MAJOR_SPORTS_CONCERT_THRESHOLD" -> BusinessRules.MAJOR_SPORTS_CONCERT_THRESHOLD,

      // Event size categorization (small, medium, large, massive)
      "EVENT_SMALL_MAX" -> BusinessRules.EVENT_SMALL_MAX,
      "EVENT_MEDIUM_MAX" -> BusinessRules.EVENT_MEDIUM_MAX,
      "EVENT_LARGE_MAX" -> BusinessRules.EVENT_LARGE_MAX,

      // Event duration categorization (brief, moderate, extended, multi-day)
      "DURATION_BRIEF_MAX" -> BusinessRules.DURATION_BRIEF_MAX,
      "DURATION_MODERATE_MAX" -> BusinessRules.DURATION_MODERATE_MAX,
      "DURATION_EXTENDED_MAX" -> BusinessRules.DURATION_EXTENDED_MAX
    )
  }

  override protected def getDeduplicationColumns(): Array[String] = {
    Array("timestamp", "location_id", "event_type", "start_time")
  }
}

object EventsCleanedJob {
  private val logger = LoggerFactory.getLogger(this.getClass)

  def main(args: Array[String]): Unit = {
    logger.info("========================================")
    logger.info("Silver Layer: Events Cleaned Job")
    logger.info("========================================")

    try {
      val config = new StreamConfig()
      val job = new EventsCleanedJob(config)
      job.run()

    } catch {
      case e: Exception =>
        logger.error("Failed to run EventsCleanedJob", e)
        System.exit(1)
    }
  }
}
